### PR TITLE
fixes #2444 - locale selector in user account

### PR DESCRIPTION
**Pull request by lzap**
Selector in the user new/edit page. Works for both admin and non-admin
users. By default it does show language country code (e.g. "es" for
Spanish), but once translators translate "locale_name" special string (it is
documented via TRANSLATORS comment) it does appear in the native name (e.g.
Čeština).

The default (empty) value is "Browser Locale" which detects automatically.

I am also fixing one line - in the new user action, Rails was giving Can't
mass-assign protected attributes: admin error - I took admin field out of
the mass assignment.

### CODE REVIEW

**Comment by ohadlevy** [comment]
@abenari please review - thanks!

**Comment by abenari** [comment]
@lzap, do we need the foreman.po files in this commit?

**Comment by lzap** [comment]
Yes, I added "locale_name" new string. And then the extraction extracted several other strings - not sure who did not do that.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -61,6 +61,12 @@ def available_sso
     @available_sso ||= SSO.get_available(self)
   end
 
+  # This filter is called before FastGettext set_gettext_locale and sets user-defined locale
+  # from db. It must be called after require_login.
+  def set_gettext_locale_db
+    params[:locale] = User.current.locale if User.current and not User.current.locale.blank? and not params[:locale]
```
Comment (ohadlevy) [inline app/controllers/application_controller.rb]: maybe use 

``` ruby
params[:locale] ||= User.current.try(:locale)
```

Comment (ohadlevy) [inline app/controllers/application_controller.rb]: I would also add to the users controller to skip this method, e.g. at https://github.com/lzap/foreman/blob/f192dce74ccde93f9664a36c4a6de634d0ca0fdd/app/controllers/users_controller.rb#L6

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -31,7 +31,8 @@ def new
   end
 
   def create
-    @user = User.new(params[:user]){|u| u.admin = params[:user][:admin] }
+    admin = params[:user].delete :admin
+    @user = User.new(params[:user]){|u| u.admin = admin }
```
Comment (ohadlevy) [inline app/controllers/users_controller.rb]: lzap, is this related to #563 - I would prefer to keep this separate PR

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -70,6 +71,9 @@ def update
     # make sure users cache are expired (assuming some permissions changed etc)
     expire_fragment("tabs_and_title_records-#{@user.id}")
     User.current.editing_self = false if editing_self
+
+    # Remove locale from the session when set to "Browser Locale" and editing self
+    session.delete(:locale) if params[:user][:locale].empty? if params[:user][:locale] and params[:id].to_i == User.current.id
```
Comment (ohadlevy) [inline app/controllers/users_controller.rb]: multiple if conditions in the same line maybe confusing? is this the same as:

``` ruby
if params[:user][:locale].blank? and params[:id].to_i == User.current.id
```

I'm not 100% sure why we need to check the params[:id], can you provide some more background?

File: config/initializers/fast_gettext.rb [hunk 4]
```diff
@@ -9,21 +9,39 @@
...
 end
 
-if Rails.env.test?
-  # in test mode we do not support i18n
+begin
+  if Rails.env.test?
+    # in test mode we do not support i18n
+    default_available_locales = []
+  else
+    default_available_locales = Dir.glob("locale/*/foreman.#{locale_type}").collect {|x| x.split('/')[1]}
```
Comment (ohadlevy) [inline config/initializers/fast_gettext.rb]: should we add Rails.root to ensure we always search in the right dir?

File: config/initializers/fast_gettext.rb [hunk 5]
```diff
@@ -9,21 +9,39 @@
...
 end
 
+# initialize fast gettext
 FastGettext.add_text_domain 'foreman',
   :path => locale_dir,
   :type => locale_type,
   :ignore_fuzzy => true,
   :report_warning => false
-FastGettext.default_available_locales = ['en'] + default_available_locales
+FastGettext.default_available_locales = ['en'] + default_available_locales.reject {|x| x == 'en'}
```
Comment (ohadlevy) [inline config/initializers/fast_gettext.rb]: why add it manually and then reject it? for order ?

**Comment by ohadlevy** [comment]
@lzap also note that the tests are failing to migrate from a clean state

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (lzap) [inline app/controllers/users_controller.rb]: I will remove these three guys as it will be fixed in there.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (lzap) [inline app/controllers/application_controller.rb]: The "try" method is ellegant, I like that. But I will have to add one extra line, because an empty string is not valid locale value. I think it may be better to use nil as "Browser Locale", but then the submitting gets more complicated.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (lzap) [inline app/controllers/users_controller.rb]: I will try to get rid of this.

File: config/initializers/fast_gettext.rb (hunk 4 above)
Comment (lzap) [inline config/initializers/fast_gettext.rb]: Right.

File: config/initializers/fast_gettext.rb (hunk 5 above)
Comment (lzap) [inline config/initializers/fast_gettext.rb]: Neeeei I was just checking if you do reviews properly... :-) Hahaah yeah, refactoring bit. Will do it properly now.

**Comment by lzap** [comment]
What's "clean state" ?

**Comment by ohadlevy** [comment]
@lzap look at the test failures from the link above

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (lzap) [inline app/controllers/users_controller.rb]: Ok the way it works in filters is it tries in this order:

```
params[:locale] || session[:locale] || cookies[:locale] ||  request.env['HTTP_ACCEPT_LANGUAGE'] || I18n.default_locale
```

In our own filter with precedes the gettext_rails_i18n filter I set params[:locale] according to user setting. This overrides everything and works well, the issue is when user have something selected in the current session and he or she puts "Browser Locale" back there. Then I must delete the session[:locale] entry because otherwise user would need to log out and log in to get rid of the old setting.

That is reason why we have this. I just badly formatted it, now with my new good friend "try" I can make it much cleaner:

```
 session.delete(:locale) if params[:user][:locale].try(:empty?) and params[:id].to_i == User.current.id
```

Note I cannot use editing_self attribute because it is never set when the current user is admin. That is the reason why I need to compare IDs.

**Comment by lzap** [comment]
Ah right, got it. Okay pushed, rebased, all fixed. [test]

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -70,6 +70,9 @@ def update
     # make sure users cache are expired (assuming some permissions changed etc)
     expire_fragment("tabs_and_title_records-#{@user.id}")
     User.current.editing_self = false if editing_self
+
+    # Remove locale from the session when set to "Browser Locale" and editing self
+    session.delete(:locale) if params[:user][:locale].try(:empty?) and params[:id].to_i == User.current.id
   end
```
Comment (ohadlevy) [inline app/controllers/users_controller.rb:77]: nil.empty? works too, it wont work if params[:user] is nil

File: config/initializers/fast_gettext.rb [hunk 7]
```diff
@@ -9,21 +9,39 @@
...
+begin
+  if Rails.env.test?
+    # in test mode we do not support i18n
+    FastGettext.default_available_locales = ['en']
+  else
+    FastGettext.default_available_locales = 
+      Dir.glob("#{Rails.root}/locale/*/foreman.#{locale_type}").collect {|x| x.split('/')[-2]}
+  end
+rescue Exception => e
+  Rails.log.warn "Unable to set avaiable locales: #{e}"
```
Comment (domcleal) [inline config/initializers/fast_gettext.rb]: typo here, "available"

**Comment by domcleal** [comment]
After logging out, I'm finding the login page is reverting to German.  I think you need to do this, so that the locale for the request is always initialised:

<pre>diff --git a/app/controllers/users_controller.rb b/app/controllers/users_controller.rb
index 8c72c4e..7ae4e81 100644
--- a/app/controllers/users_controller.rb
+++ b/app/controllers/users_controller.rb
@@ -3,7 +3,7 @@ class UsersController < ApplicationController
 
   before_filter :find_user, :only => [:edit, :update, :destroy]
   skip_before_filter :require_mail, :only => [:edit, :update, :logout]
-  skip_before_filter :require_login, :authorize, :session_expiry, :update_activity_time, :set_taxonomy, :set_gettext_locale, :set_gettext_locale_db, :only => [:login, :logout]
+  skip_before_filter :require_login, :authorize, :session_expiry, :update_activity_time, :set_taxonomy, :set_gettext_locale_db, :only => [:login, :logout]
   after_filter       :update_activity_time, :only => :login
 
   attr_accessor :editing_self
</pre>

**Comment by lzap** [comment]
Reproduced - only when I log out right after user edit form is submitted. Your patch works great, thanks! Rebasing.

**Comment by domcleal** [comment]
:+1: working reliably for me now, thanks.

**Comment by ohadlevy** [comment]
merged - thanks
